src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let copy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case "addProperties":
+        Object.assign(copy, action.extraData);
+        break;
+
+      case "removeProperties":
+        action.keysToRemove.forEach((key) => delete copy[key]);
+        break;
+
+      case "clear":
+        copy = {};
+        break;
+
+      default:
+        throw new Error(
+          action.type
+            ? `${action.type} is not supported`
+            : `Action is ${action.type}`
+        );
```

Comment:
you refer to action type anyway here, even if it is undefined. So basically you do the same thing in both cases